### PR TITLE
monit: fix module detection of monitored process state

### PR DESCRIPTION
**Pull request by snopoke**
##### SUMMARY
The monit module has been broken since at least [2017-03-06](https://bitbucket.org/tildeslash/monit/commits/8546c07ea1fe04e058567e668b7d64ebc2ee1a4f) (and probably before) when the command output that it uses to get the process status changed.

This change to monit's CLI output prevents the module from detecting certain states of the process, notably the 'running' state. This means that certain module actions do not work as exected.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
monit

##### ADDITIONAL INFORMATION
See https://github.com/ansible-collections/community.general/issues/1106

### CODE REVIEW

File: tests/integration/targets/monit/files/httpd_echo.py
Comment (snopoke) [inline tests/integration/targets/monit/files/httpd_echo.py:1]: I'm not at all attached to these python files (httpd_echo.py and daemon.py). I just needed a process that monit could monitor and wanted something very simple and without any other dependencies.

**Comment by snopoke** [comment]
In the build output I see `WARNING: All targets skipped.` for all the integration jobs. Is that because the integration is marked as `destructive`? Is there a recommended way to deal with that?

**Comment by aminvakil** [comment]
I've tested the integration tests locally and there were some errors, maybe this can help you.

**Comment by aminvakil** [comment]
Ubuntu 16.04:
```
$ ansible-test integration monit --docker ubuntu1604
WARNING: Cannot check pip requirements for conflicts because "pip check" is not supported.
WARNING: A Python 2.7 interpreter was found at "/usr/bin/python2.7", yet a matching pip was not found.
WARNING: A Python 3 interpreter was found at "/usr/bin/python3", yet a matching pip was not found.
WARNING: A Python 3.5 interpreter was found at "/usr/bin/python3.5", yet a matching pip was not found.
Running monit integration test role
[WARNING]: running playbook inside collection community.general

PLAY [testhost] ****************************************************************

TASK [Gathering Facts] *********************************************************
ok: [testhost]

TASK [setup_pkg_mgr : set_fact] ************************************************
skipping: [testhost]

TASK [setup_pkg_mgr : set_fact] ************************************************
skipping: [testhost]

TASK [install monit] ***********************************************************
[WARNING]: Updating cache and auto-installing missing dependency: python-apt
changed: [testhost]

TASK [monit config] ************************************************************
changed: [testhost]

TASK [process monit config] ****************************************************
changed: [testhost]

TASK [monit : copy process file] ***********************************************
changed: [testhost] => (item=daemon.py)
changed: [testhost] => (item=httpd_echo.py)

TASK [restart monit] ***********************************************************
changed: [testhost]

TASK [monit : include_tasks] ***************************************************
included: /root/ansible_collections/community/general/tests/output/.tmp/integration/monit-O10t8O-ÅÑŚÌβŁÈ/tests/integration/targets/monit/tasks/test.yml for testhost

TASK [monit : verify initial state (up)] ***************************************
fatal: [testhost]: FAILED! => {"changed": true, "cmd": ["curl", "-sf", "http://localhost:8082/hello"], "delta": "0:00:00.037480", "end": "2020-10-16 18:39:39.652410", "msg": "non-zero return code", "rc": 7, "start": "2020-10-16 18:39:39.614930", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [stop monit] **************************************************************
changed: [testhost]

TASK [uninstall monit] *********************************************************
changed: [testhost]

PLAY RECAP *********************************************************************
testhost                   : ok=9    changed=7    unreachable=0    failed=1    skipped=2    rescued=0    ignored=0   

NOTICE: To resume at this test target, use the option: --start-at monit
ERROR: Command "ansible-playbook monit-qoE_mX.yml -i inventory" returned exit status 2.
ERROR: Command "docker exec 118855014dedc00a4fc376f3606d1b9bb6751c9eb24523ffc3d4273fe75bbde3 /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/general LC_ALL=en_US.UTF-8 /usr/bin/python2.7 /root/ansible/bin/ansible-test integration monit --metadata tests/output/.tmp/metadata-98nwqxys.json --truncate 162 --redact --color yes --requirements --allow-destructive" returned exit status 1.
```

**Comment by aminvakil** [comment]
CentOS 7:
```
$ ansible-test integration monit --docker centos7
Collecting idna<2.6,>=2.5 (from -c /root/ansible/test/lib/ansible_test/_data/requirements/constraints.txt (line 16))
  Downloading https://files.pythonhosted.org/packages/11/7d/9bbbd7bb35f34b0169542487d2a8859e44306bb2e6a4455d491800a5621f/idna-2.5-py2.py3-none-any.whl (55kB)
Installing collected packages: idna
  Found existing installation: idna 2.4
    Uninstalling idna-2.4:
      Successfully uninstalled idna-2.4
Successfully installed idna-2.5
WARNING: Cannot check pip requirements for conflicts because "pip check" is not supported.
Running monit integration test role
[WARNING]: running playbook inside collection community.general

PLAY [testhost] ****************************************************************

TASK [Gathering Facts] *********************************************************
ok: [testhost]

TASK [setup_pkg_mgr : set_fact] ************************************************
skipping: [testhost]

TASK [setup_pkg_mgr : set_fact] ************************************************
skipping: [testhost]

TASK [install monit] ***********************************************************
changed: [testhost]

TASK [monit config] ************************************************************
fatal: [testhost]: FAILED! => {"changed": false, "checksum": "2fecf71f8c3fec6a829760aa45e079fd2dc0767c", "msg": "Destination directory /etc/monit does not exist"}

TASK [stop monit] **************************************************************
ok: [testhost]

TASK [uninstall monit] *********************************************************
changed: [testhost]

PLAY RECAP *********************************************************************
testhost                   : ok=4    changed=2    unreachable=0    failed=1    skipped=2    rescued=0    ignored=0   

NOTICE: To resume at this test target, use the option: --start-at monit
ERROR: Command "ansible-playbook monit-d83fHL.yml -i inventory" returned exit status 2.
ERROR: Command "docker exec f77d916ba1860883b3c8ee6b341e063ca7c0478328a4b3027da8de10856e43a5 /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/general LC_ALL=en_US.UTF-8 /usr/bin/python2.7 /root/ansible/bin/ansible-test integration monit --metadata tests/output/.tmp/metadata-m0u87ehw.json --truncate 162 --redact --color yes --requirements --allow-destructive" returned exit status 1.
```

**Comment by snopoke** [comment]
@aminvakil thanks, I'll look into those. Is there a way to run the tests on all the envs?

**Comment by aminvakil** [comment]
I can't find something in https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html#tests-in-docker-containers regarding testing on all images, but maybe something like this? :)
`for image in ubuntu1604 centos7; do ansible-test integration monit --docker $image; done`

Can I ask why the integration test has been marked as `destructive` if you are stopping the services and uninstalling packages which the test has been started and installed?

**Comment by felixfontein** [comment]
@aminvakil if the user already had that installed before running the test, and didn't run it in a container / VM, it is uninstalled afterwards. So it is destructive :)

File: tests/integration/targets/monit/aliases
Comment (felixfontein) [inline tests/integration/targets/monit/aliases:1]: You need to add
```suggestion
shippable/posix/group2
destructive
```
so that it is actually run in CI. If they are not assigned to any group, they will be skipped.

Comment (felixfontein) [reply tests/integration/targets/monit/aliases:1]: You probably want to skip some of the platforms, like add things like
```
skip/osx
skip/macos
skip/freebsd
skip/aix
```
(or a subset of that).

**Comment by felixfontein** [comment]
BTW, since you are basically rewriting the module, could you please remove its mentions from `tests/sanity/ignore-*.txt` and fix the errors? (I.e. add `type: ...` in `DOCUMENTATION` or the options.)

Since you're doing a rewrite, you could also add yourself to the authors list (simply move the current author to the first list element, and put yourself as the second list element in the same format).

**Review by Andersson007 — COMMENTED** [review]
@snopoke thanks for fixing this!
please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

File: tests/integration/targets/monit/aliases
Comment (snopoke) [reply tests/integration/targets/monit/aliases:1]: ed4ce0f6

**Comment by snopoke** [comment]
Thanks for all the feedback. I was able to run it successfully on `fedora31`, `opensuse15`, `centos7`, `ubuntu1804`

**Comment by snopoke** [comment]
Build is failing. I will look into the errors.

**Comment by aminvakil** [comment]
Adding `type: str` to module documentation fixes the sanity tests.
```
   name:
     description:
       - The name of the I(monit) program/process to manage
+    type: str
     required: true
   state:
     description:
       - The state of service
+    type: str
     required: true
     choices: [ "present", "started", "stopped", "restarted", "monitored", "unmonitored", "reloaded" ]
```
It seems that installing `monit` on CentOS 6, 7, 8 needs `EPEL` repo.

So I guess it needs to be installed before installing `monit` package itself and if it was installed during in `monit` integration test, it should be removed at the end.

**Comment by felixfontein** [comment]
About installing EPEL: grep for `epel` in tests/integration/targets, you'll fine some hints on how to do that. It looks like it's essentially:
```.yaml
- name: Install EPEL repository (RHEL only)
  include_role:
    name: setup_epel
  when: ansible_distribution == 'RedHat'
```

**Comment by snopoke** [comment]
thanks @felixfontein, I also figured that out. I'm just working on the ubuntu1604 build now which installs an old version of monit.

**Comment by felixfontein** [comment]
The only sanity error left:
```
plugins/modules/monitoring/monit.py:65:1: E302: expected 2 blank lines, found 1
```

File: tests/integration/targets/monit/tasks/main.yml
Comment (felixfontein) [inline tests/integration/targets/monit/tasks/main.yml:69]: I would remove this task.

Comment (snopoke) [reply tests/integration/targets/monit/tasks/main.yml]: f2bd0570

File: changelogs/fragments/1107-monit-fix-status-check.yml
Comment (Andersson007) [inline changelogs/fragments/1107-monit-fix-status-check.yml:2]:
```suggestion
  - monit - fix modules ability to determine the current state of the monitored process (https://github.com/ansible-collections/community.general/pull/1107).
```
in accordance to the doc

File: plugins/modules/monitoring/monit.py
Comment (Andersson007) [inline plugins/modules/monitoring/monit.py:16]:
```suggestion
    - Manage the state of a program monitored via I(Monit).
```
`short_description` shouldn't have a final dot but `description` must have

File: plugins/modules/monitoring/monit.py
Comment (Andersson007) [inline plugins/modules/monitoring/monit.py:20]:
```suggestion
      - The name of the I(monit) program/process to manage.
```

File: plugins/modules/monitoring/monit.py
Comment (Andersson007) [inline plugins/modules/monitoring/monit.py:25]:
```suggestion
      - The state of service.
```

**Review by Andersson007 — COMMENTED** [review]
my 50 cent:)

**Comment by Andersson007** [comment]
@felixfontein we definitely should add @snopoke as a maintainer of the module after merging;)

**Comment by felixfontein** [comment]
@Andersson007 assuming he wants. @snopoke what do you think, are you interested in becoming a maintainer for this module? You will get automatically CC'ed for every issue and PR, and your vote for/against changes will count more. (Also you can stop down from being a maintainer at any point.)

**Comment by snopoke** [comment]
@Andersson007 @felixfontein I'm happy to be the maintainer for now.

**Comment by felixfontein** [comment]
recheck

File: tests/integration/targets/monit/files/daemon.py
Comment (felixfontein) [inline tests/integration/targets/monit/files/daemon.py:2]: I think this file needs a copyright header. I assume it is GPLv3+ licensed?

File: tests/integration/targets/monit/files/httpd_echo.py
Comment (felixfontein) [inline tests/integration/targets/monit/files/httpd_echo.py:2]: Same here. Where does this file comes from? It should have copyright and license header.

**Review by felixfontein — COMMENTED** [review]
The tests seem to not run smooth in all cases:
https://app.shippable.com/github/ansible-collections/community.general/runs/4536/51/tests
https://app.shippable.com/github/ansible-collections/community.general/runs/4562/51/tests
https://app.shippable.com/github/ansible-collections/community.general/runs/4584/51/tests
(the latter two were re-runs of the first - 'unstable' means that when the test was run another time, it suddenly succeeded, and 'failed' means that the test failed two times in a row)

File: tests/integration/targets/monit/files/daemon.py
Comment (snopoke) [reply tests/integration/targets/monit/files/daemon.py:2]: I've no idea how to license this. It was written by an anonymous source: https://www.jejik.com/articles/2007/02/a_simple_unix_linux_daemon_in_python/

I can try the `python-daemon` library instead and see if it works on all the builds.

File: tests/integration/targets/monit/files/httpd_echo.py
Comment (snopoke) [reply tests/integration/targets/monit/files/httpd_echo.py:2]: I wrote this, I'll put a GPL v3 on it.

File: tests/integration/targets/monit/files/daemon.py
Comment (snopoke) [reply tests/integration/targets/monit/files/daemon.py:2]: I've got some builds working. I had to skip python2.6. I'll push now and see how the build does. I've also refactored the tests a bit and tried to make it more robust.

**Comment by snopoke** [comment]
Unstable on centos8 this time: https://app.shippable.com/github/ansible-collections/community.general/runs/4608/121/tests

I'll try to repo locally.

**Comment by felixfontein** [comment]
@snopoke thats that sometimes fail are ok, but if they fail too often, it has a pretty bad impact on CI, so we try to avoid that. Thanks for trying to fix this! :)

**Comment by snopoke** [comment]
OK, I've been running this in a loop and I've only hit that issue once out of > 200 attempts. I was hoping to hit it again after adding some more debug tasks but no such luck.

I have made a small change which I hope will make the failure less likely.

**Comment by snopoke** [comment]
Tests still unstable. I'm going to add some debugging in to try and get some more info about why it's failing.

**Comment by felixfontein** [comment]
If you simply want to re-run tests, you can also close+reopen the PR.

**Comment by snopoke** [comment]
Based on the debug output it seems like the monit daemon itself is sometimes taking a long time to initialize (or get's stuck in that state). I've modified the test to retry the task a few times.

**Comment by snopoke** [comment]
Close / reopen to re-trigger build.

**Comment by snopoke** [comment]
@Andersson007 @felixfontein I've remove the flaky task from the test suite - the build has now run twice in a row with no failures. I think this is ready to go unless there's any additional feedback.

**Comment by felixfontein** [comment]
I'll re-run the tests a couple of times more (I can directly trigger a shippable re-run), just to be on the safe side.

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
I triggered the tests three times in a row, and they always passed. \o/

**Comment by felixfontein** [comment]
@Andersson007 @aminvakil do you have more comments, or should we merge?

**Comment by aminvakil** [comment]
@felixfontein I was just thinking right now about it:) Is it OK if @snopoke changes the test and put `retry until success` in the sometimes failing task?

**Comment by felixfontein** [comment]
@aminvakil do you mean unlimited retries instead of max. 5?

**Comment by aminvakil** [comment]
@felixfontein I should have get my facts right before saying this, I haven't seen the failing test, it's already there.
No, I didn't mean unlimited retires, I meant exactly the way it is now.

**Comment by snopoke** [comment]
I did already try retires and also putting in a `block` with a `rescue` (which still failed the build). It would seem that once the process got into that state it would not recover so no amount of retries would help. In the end I took a different approach with the tests to avoid hitting that issue. Note that the issue isn't with the Ansible module but with the underlying service itself.

**Review by Andersson007 — APPROVED** [review]

**Comment by felixfontein** [comment]
@snopoke thanks a lot for working on this module, especially adding tests!
@aminvakil @Andersson007 thanks a lot for testing and reviewing!

**Comment by Andersson007** [comment]
@felixfontein don't forget to add @snopoke as a maintainer;)
thanks everyone!

**Pull request merged**
